Implement correct export/print stages and dialog behavior

creviera opened the pull request:
# Description

* closes #642
* Towards https://github.com/freedomofpress/securedrop-client/issues/549

A lot of messaging has been done as a group during meetings and in Slack. We still need to finalize the messaging so this can still be considered a rough draft but closer to what we want.

* Resolves https://github.com/freedomofpress/securedrop-client/issues/557

In order to do this in Qubes I had to set the window flag to `Qt.Widget` (some window flags in Qubes do not work unless you tell Qt to bypass the window manager, which is a dangerous thing to do). Since the dialog is a widget, the user can continue to interact with the client while the dialog is open. In order to avoid a race condition, I made it so only one print or export dialog can be opened at a time. Since the dialog cannot be moved, it is very unlikely that the user will try to interact with the client while it is open, but we need to make sure nothing bad happens if they do during test (see Test Plan).

This change also required adding close and cancel buttons to close the dialog since it can no longer be closed by clicking outside of it. Since the cancel button was added, I added a continue button next to it so we could see what the buttons look like together. 

* Resolves https://github.com/freedomofpress/securedrop-client/issues/618

This should have been a separate PR and can be factored out, but will take a bit of effort, so I want to offer a code walk-through first in case that helps. The speedbump was added for both print and export, which was one more reason to create a FramelessDialog class that they could both derive from. The FramelessDialog class sets the window flags and creates the layout and styles that both print and export share. The layout looks like this: 
![Screenshot from 2020-01-21 01-04-35](https://user-images.githubusercontent.com/4522213/72790266-2bb6ae80-3bea-11ea-8adf-aa6f27891274.png)

* Resolves updating the UI, see https://app.zeplin.io/project/5c807ea562f734bd2756b243/screen/5ddc3b8b9d9aab94be0542d5 for reference, which has gone through many iterations

* Added filename to PrintDialog so that "Preparing to print" -> "Preparing to print: filename.pdf"
* Added "Export successful" message since we removed it from the Export VM. I created a [followup issue](https://github.com/freedomofpress/securedrop-client/issues/713) to make the color green and icon image to a checkmark.
* Added show/hide password to passphrase form (using the same eye icon as we use in the login dialog)

# Test Plan

1. Click outside of dialog for both print and export to verify it no longer closes the dialog
2. Make sure you can only open one dialog at a time, either a single print dialog or a single export dialog
3. Resize the window and see that the dialog continues to be centered with the application
4. Click on "Print" and see that filename now shows up in preparation window
5. Click on "Cancel" to verify that it closes the dialog for both print and export
6. Test show/hide password
7. See "Export successful" message when the export succeeds
8. See dialog close when the Export VM begins opening the 3rd party printer application
9. Make sure you can still export and print documents and test all the error cases: (1) bad passphrase, (2) no usb device connected, (3) multiple devices connected, (4) unencrypted usb device connected

creviera commented:
Also verify that "Continue" is now the default selection

creviera commented:
@redshiftzero heads up: a PR for [this issue](https://github.com/freedomofpress/securedrop-export/issues/23) will need to be merged first and then i'll need update this pr to use the new printer pre-flight check so that the client dialog can close when X Printing Panel is in use.

kushaldas commented on securedrop_client/gui/widgets.py:
```diff
@@ -1968,23 +1966,41 @@ class PrintDialog(QDialog):
+            Qt.Dialog |
+            Qt.FramelessWindowHint |
+            Qt.WindowStaysOnTopHint |
+            Qt.X11BypassWindowManagerHint)
```
Nice set of defaults.

eloquence commented:
This is blocked on https://github.com/freedomofpress/securedrop-export/issues/23 (WIP in a branch).

creviera commented:
A lot of rebasing since this is getting to be an old PR and updates based on in-person feedback during design reviews. Sorry for the size, I updated the PR description and mentioned that I can split this into two PRs but would be happy to do a code-walkthrough.

 **BEFORE (first step)**

![Screenshot from 2020-01-21 01-13-05](https://user-images.githubusercontent.com/4522213/72791039-800e5e00-3beb-11ea-9644-16d7fbafaee2.png)
![Screenshot from 2020-01-21 01-12-25](https://user-images.githubusercontent.com/4522213/72791040-800e5e00-3beb-11ea-93b9-3c6c0d67dbd9.png)

**AFTER (first step)**

![Screenshot from 2020-01-21 00-24-33](https://user-images.githubusercontent.com/4522213/72791071-8c92b680-3beb-11ea-95bd-591d1e2e1d96.png)
![Screenshot from 2020-01-21 00-25-25](https://user-images.githubusercontent.com/4522213/72791065-8997c600-3beb-11ea-81a7-f387234e7b74.png)

**AFTER (additional steps, export for example)**
![Screenshot from 2020-01-21 00-33-18](https://user-images.githubusercontent.com/4522213/72792242-89002f00-3bed-11ea-84b2-983805cfbff1.png)
![Screenshot from 2020-01-21 01-31-41](https://user-images.githubusercontent.com/4522213/72792590-25c2cc80-3bee-11ea-93c8-d2294d24f3da.png)
![Screenshot from 2020-01-21 01-32-23](https://user-images.githubusercontent.com/4522213/72792589-25c2cc80-3bee-11ea-9232-fa322313e49f.png)
![Screenshot from 2020-01-21 01-32-37](https://user-images.githubusercontent.com/4522213/72792588-252a3600-3bee-11ea-9e55-f069a1b18ee2.png)
![Screenshot from 2020-01-21 01-33-23](https://user-images.githubusercontent.com/4522213/72792587-252a3600-3bee-11ea-8aea-1c63a92a97bf.png)
![Screenshot from 2020-01-21 01-36-06](https://user-images.githubusercontent.com/4522213/72792779-6ae6fe80-3bee-11ea-92bf-e6c16fce18af.png)

This is ready for review!

creviera commented on securedrop_client/gui/widgets.py:
```diff
@@ -2007,351 +2009,538 @@ def _on_left_click(self):
-        self.setWindowModality(Qt.WindowModal)
+        self.setWindowFlags(Qt.Widget)
 
+        # Set drop shadow effect
```
this helps make the widget look like a dialog

creviera commented on securedrop_client/gui/widgets.py:
```diff
@@ -1971,11 +1968,11 @@ def _on_export_clicked(self):
+        if not self.dialog_in_progress:
+            self.dialog_in_progress = True
+            dialog = ExportDialog(self.controller, self.file.uuid, self.file.original_filename)
+            dialog.dialog_closing.connect(self._unset_dialog_in_progress)
```
this is how we know that we can unset `dialog_in_progress`

creviera commented on securedrop_client/gui/widgets.py:
```diff
@@ -2007,351 +2009,538 @@ def _on_left_click(self):
 
-        self.controller = controller
-        self.file_uuid = file_uuid
+    dialog_closing = pyqtSignal()
```
used for making sure only one dialog is open at a time

creviera commented:
> 3. Resize the window and see that the dialog continues to be centered with the application

I lied here in the test plan. What I meant to say is that the dialog recenters after the window resizes during each step.

redshiftzero commented on securedrop_client/gui/widgets.py:
```diff
@@ -2007,351 +2009,538 @@ def _on_left_click(self):
+        self.header_icon = SvgLabel('blank.svg', svg_size=QSize(64, 64))
+        self.header_icon.setObjectName('header_icon')
+        self.header_icon.setFixedWidth(80)
+        self.header = QLabel()
```
in the child classes here we're using this widget to display the filename which is user-controlled input, so this should be SecureQLabel

creviera replied:
`SecureQLabel` gets rid of the formatting needed to make the header look like the zeplin spec unfortunately. The header would need to be updated to be two SecureQLabels because there are two different styles in the same header: one for the "Task name" the other for the filename.

redshiftzero commented on securedrop_client/gui/widgets.py:
```diff
@@ -2007,351 +2009,538 @@ def _on_left_click(self):
+            'Documents submitted by sources may contain information or hidden metadata that '
+            'identifies who they are. To protect your sources, please consider redacting documents '
+            'before working with them on network-connected computers.')
+        self.exporting_message = _('Exporting: {}'.format(self.file_name))
```
is this unused? I was looking through these attributes to trace where `file_name` gets used to check for potential injection issues and can't find where this attribute `exporting_message` is used

creviera replied:
ah, good find. I can remove this line.

redshiftzero replied:
`exporting_message` still unused?

creviera commented:
> Let me know if you also see the above behavior. I’m thinking this is due to the sequence of .show() and .hide() that are called on these widgets but I haven't traced yet in detail.

Ah, yes the ordering of hiding and showing for widgets that need to adjust size actually matters a lot. I think you are right and things were looking better before i moved all the show calls before the hide calls. Let me see if going back to the old ordering of show and hide calls fixes the issue. Somewhere between ordering of show/hide calls and creating unique widgets for each message rather than using setText on the body and error_details section is probably the answer...

creviera commented:
closes https://github.com/freedomofpress/securedrop-client/issues/642

creviera commented on securedrop_client/gui/widgets.py:
```diff
@@ -1943,6 +1936,10 @@ def __init__(
         file_ready_signal.connect(self._on_file_downloaded, type=Qt.QueuedConnection)
 
+        # Make sure we only allow one export or print operation at a time (workaround for frameless
+        # modals not working as expected on QubesOS)
```
the flag Qt.FramelessWindowHint doesn't seem to work in Qubes, so I had to set the Qt.Widget flag in order to get rid of the frame and then add a shadow and prevent clicking on "Print" and "Export" opening another dialog

ninavizz commented:
@creviera Have you been waiting for me to review or submit changes before Merging this?

eloquence commented:
For 2/5-2/19 sprint, @creviera will rebase/retest but we're not aiming for review and merge in this sprint yet (focusing on bugfixes), just keeping the PR in mint condition so it can hopefully get over the finish line in the final pre-release sprint.

eloquence commented:
This will also resolve #556, correct? With #713 as follow-up?

creviera commented on securedrop_client/gui/widgets.py:
```diff
@@ -2007,351 +2009,538 @@ def _on_left_click(self):
+        self.error_details.show()
+        self.passphrase_form.show()
+        self.adjustSize()
+        self.center_dialog()
```
ah, i was just reviewing this PR, and it looks like the dialog is actually recentered in qubes when the app resizes (not just when you click continue), perhaps when this is ready for another review, you can test this again in qubes to verify this behavior.

creviera commented:
This is ready for review and should be tested in Qubes. Two known issues: (1) the continue button doesn't change to a slightly light color (reducing opacity) after it's been clicked on yet, and (2) the dialog doesn't recenter when you resize the client window, which i will file followup issues for.

redshiftzero commented:
I'm leaving my blocking review here until I check [one issue](https://github.com/freedomofpress/securedrop-client/pull/666#discussion_r369266529) but in the meantime - @kushaldas want to take a spin through the functionality & diff here?

creviera commented:
I forgot about addressing that comment, thanks @redshiftzero 

Instead of splitting the header into two separate widgets (most importantly a QLabel for the filename), we could sanitize the filename input before adding it to the one header widget, as discussed. I'll update this so it'll be ready for kushal's review.

creviera commented:
> I'm leaving my blocking review here until I check [one issue](https://github.com/freedomofpress/securedrop-client/pull/666#discussion_r369266529) but in the meantime - @kushaldas want to take a spin through the functionality & diff here?

just added a commit that addresses this issue (also added a couple tests to ensure filename is sanitized using our standard sensitization method when the dialogs are created)

creviera commented:
> * :x: When I exported a file, it created 3 exported directories in the device.

It looks like this is a bug in the continue button signal/slot wiring for _both_ print and export operations. I don't know why this is happening, but here's the simplest STR I could come up with:

1. Do not connect a printer
2. run `LOGLEVEL=debug python -m securedrop_client`
3. click on print button
4. before you click continue button after reading the printer safety tips, tail the logs in both `sd-devices` for securedrop-export and `sd-app` for securedrop-client
5. click continue
6. see `ERROR_PRINTER_NOT_FOUND` log
7. click continue again
8. see multiple `ERROR_PRINTER_NOT_FOUND` logs

After adding extra log lines while debugging, i see that the problem is coming from the continue button. Either one click is somehow being emitted as multiple clicks or there is something wrong with how I'm connecting the continue button to the slot: `self.continue_button.clicked.connect(self._run_preflight)`. 

I'll do some more debugging and perhaps we can discuss this issue further tomorrow. (Nice find)

creviera commented:
> * :x: If I connect an un-encrypted USB device, there is no way to know what is going on when I click on "continue" in the export dialog

## No device

This is what I'm seeing is when no drive is attached (the screen doesn't change if you click continue without attaching a device):
![no-device-attached-after](https://user-images.githubusercontent.com/4522213/74996286-05ae4500-5408-11ea-90a2-5b027f2253eb.png)

When you click the continue button we don't show an error message in red, which we should, but I'd like to point out that we don't on `master` either. If there's time tomorrow, I could squeeze this into this PR or we could do it as a followup. Here's what you'll see on `master` with nothing changing when you click continue with no device attached:

![no-device-attached-before](https://user-images.githubusercontent.com/4522213/74996597-dc41e900-5408-11ea-9bd5-da5a49ba260f.png)

## Non-luks encrypted device

This is what I'm seeing is when you click continue and the device attached is not luks encrypted:

![not-luks-encrtyped-after](https://user-images.githubusercontent.com/4522213/74996760-4f4b5f80-5409-11ea-8d20-a4a745cfed1b.png)

Which is an improvement to what we do on `master`:

![not-luks-encrypted-before](https://user-images.githubusercontent.com/4522213/74996779-5bcfb800-5409-11ea-8f1f-6c794d5a1d33.png)

creviera commented:
> It looks like this is a bug in the continue button signal/slot wiring for both print and export operations.

Found the bug: rewiring a button signal requires calling disconnect on that signal in order to break previous connections

I fixed this issue and it can be tested both by running through my STR to test print here: https://github.com/freedomofpress/securedrop-client/pull/666#issuecomment-589458297 and by running through this procedure to test export:

1. go through the export flow until you successfully transfer a document to a luks-encrypted drive
2. verify only one copy of the document is on your thumb drive

redshiftzero commented:
Re-reviewed diff, my comments were addressed except [one nit](https://github.com/freedomofpress/securedrop-client/pull/666#discussion_r383352364). Testing is looking great, one issue discovered: 

✅1. Click outside of dialog for both print and export to verify it no longer closes the dialog
❌2. Make sure you can only open one dialog at a time, either a single print dialog or a single export dialog - I can open an export dialog for doc A, then click on print _for another document doc B_, and a print dialog for doc B appears on top of the export dialog for doc A. 
✅3. Resize the window and see that the dialog continues to be centered with the application - true when I click continue
✅4. Click on "Print" and see that filename now shows up in preparation window
✅5. Click on "Cancel" to verify that it closes the dialog for both print and export
✅6. Test show/hide password
✅7. See "Export successful" message when the export succeeds
✅8. See dialog close when the Export VM begins opening the 3rd party printer application
✅9. Make sure you can still export and print documents and test all the error cases: (1) bad passphrase, (2) no usb device connected, (3) multiple devices connected, (4) unencrypted usb device connected - tested no printer connected case, tested printer connected & success case, printer error case, tested USB not connected case, tested USB connected but not encrypted case, tested USB export success case, tested USB bad passphrase case. All works as expected.
